### PR TITLE
BaSD-MS-38: Create Psychologists View

### PR DESCRIPTION
**Pull request by clementinob**
### Create Psychologists View
_Components created:_

- **CreateBtn**
- **DeleteBtn**
- **Form**
- **Input**
- **Modal**
- **Psychologists**

_CRUD functionality:_
**List of Psychologists** on index of entity.
**Create Psychologist** on index of entity.
**Edit Psychologist** (clicking in the corresponding row)
**Delete Psychologist** (clicking in the corresponding button at the end of the row)

### CODE REVIEW

File: src/Components/Psychologists/Form/index.js [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    options.method = 'PUT';
+    url = `${process.env.REACT_APP_API}/psychologists/${psychologistId}`;
+  }
+
+  const onSubmit = (event) => {
+    event.preventDefault();
+
+    if (psychologistId) {
+      options.method = 'PUT';
+      url = `${process.env.REACT_APP_API}/psychologists/${psychologistId}`;
```
Comment (ravegliamartin) [inline src/Components/Psychologists/Form/index.js:53]: You're already assigning method and url if adminId in line 44-45. This logic is duplicated, I think just one should remain. Perhaps, this one in the `onSubmit` makes more sense, as here in this function is were the fetch is done.

File: src/Components/Psychologists/Form/index.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+      url = `${process.env.REACT_APP_API}/psychologists`;
+    }
+    fetch(url, options)
+      .then((response) => {
+        if (response.status !== 200 && response.status !== 201) {
+          return response.json().then(({ ErrMessage }) => {
+            throw new Error(ErrMessage);
+          });
+        }
+        return response.json();
```
Comment (ravegliamartin) [inline src/Components/Psychologists/Form/index.js:65]: Why returning a response.json() if you're doing nothing with that response? Maybe, this piece of code might be refactor as,

```
   fetch(url, options)
      .then((response) => {
        if (response.status !== 200 && response.status !== 201) {
          return response.json().then(({ ErrMessage }) => {
            throw new Error(ErrMessage);
          });
        }
         window.location.href = '/psychologists';
      })
      .catch((error) => {
        console.log(error);
      });
```

**Review by ravegliamartin — CHANGES_REQUESTED** [review]
The update function is broken but it's not your fault. 
In the server-side psychologist controller, the update function for this entity is as follow,

```
const editPsychologist = (req, res) => {
  Psychologists.findOneAndUpdate(req.params.id,
    {firstName: req.body.firstName,
    lastName: req.body.lastName,
    userName: req.body.userName,
    email: req.body.email,
    password: req.body.password}, {new : true}
  ) 
  .then((result) => {
    if (!result) {
      return res.status(400).json({
        msg:`Psychologist with id: ${req.params.id} was not found`
      })
    }
    return res.status(201).json(result)
  })
  .catch((error) => {
    return res.status(400).json(error)
  })
}
```

The condition on `findOneAndUpdate` is wrong, it is not referencing any attribute. So, the method creates a new document as it doesn't find a matching existing one. To fix this, the code on the controller should be,

```
const editPsychologist = (req, res) => {
  Psychologists.findOneAndUpdate(
    { _id: req.params.id },
    {
      firstName: req.body.firstName,
      lastName: req.body.lastName,
      userName: req.body.userName,
      email: req.body.email,
      password: req.body.password,
    },
    { new: true }
  )
    .then((result) => {
      if (!result) {
        return res.status(400).json({
          msg: `Psychologist with id: ${req.params.id} was not found`,
        });
      }
      return res.status(201).json(result);
    })
    .catch((error) => {
      return res.status(400).json(error);
    });
};
```

File: src/Components/Psychologists/Modal/modal.module.css [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+.modal button {
+  display:inline-block;
+  justify-content: center;
+  justify-content: space-around;
+  box-sizing: border-box;
+	width: 40%;
+	max-width: 580px;
+	padding: 15px;
+  margin: 80px 30px;
+
```
Comment (nicolasviola) [inline src/Components/Psychologists/Modal/modal.module.css:36]: Linea en blanco de mas

File: src/Components/Psychologists/Modal/modal.module.css [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  display:inline-block;
+  justify-content: center;
+  justify-content: space-around;
+  box-sizing: border-box;
+	width: 40%;
+	max-width: 580px;
+	padding: 15px;
+  margin: 80px 30px;
+	
+	border-top-style: none;
```
Comment (nicolasviola) [inline src/Components/Psychologists/Modal/modal.module.css:37]: Indentado

File: src/Components/Psychologists/Modal/modal.module.css [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  color: #fff;
+  text-align: center;
+}
+
+.confirmBtnModal{
+  background: #72bc8e;
+}
+
+.confirmBtnModal:hover {
+	background: #86d6a5;
```
Comment (nicolasviola) [inline src/Components/Psychologists/Modal/modal.module.css:49]: Indentado

File: src/Components/Psychologists/Modal/modal.module.css [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+.cancelBtnModal{
+  background: #646464;
+}
+
+.cancelBtnModal:hover {
+	background: #797878;
```
Comment (nicolasviola) [inline src/Components/Psychologists/Modal/modal.module.css:57]: Indentado

File: src/Components/Psychologists/index.js [hunk 7]
```diff
@@ -1,10 +1,46 @@
...
 import styles from './psychologists.module.css';
 
 function Psychologists() {
+  const [psychologists, setPsychologists] = useState([]);
+  useEffect(() => {
+    fetch(`${process.env.REACT_APP_API}/psychologists`)
+      .then((response) => response.json())
+      .then((response) => {
+        setPsychologists(response);
+      });
```
Comment (nicolasviola) [inline src/Components/Psychologists/index.js:13]: Agregar un catch. Lo mismo que te dije mas arriba

File: src/Components/Psychologists/psychologists.module.css [hunk 8]
```diff
@@ -1,6 +1,42 @@
...
\ No newline at end of file
+  justify-content: center;
+}
+
+.table{
+  width: 65%;
+  padding: 0;
+  margin: 0;
+}
+
```
Comment (nicolasviola) [inline src/Components/Psychologists/psychologists.module.css:24]: Linea en blanco de mas

**Review by nicolasviola — CHANGES_REQUESTED** [review]

File: src/Components/Psychologists/Form/index.js (hunk 2 above)
Comment (clementinob) [reply src/Components/Psychologists/Form/index.js:65]: Perfect, Rave. Thanks for pointing me that! 👍

File: src/Components/Psychologists/Form/index.js (hunk 1 above)
Comment (clementinob) [reply src/Components/Psychologists/Form/index.js:53]: Done 😄

File: src/Components/Psychologists/Form/index.js [hunk 9]
```diff
@@ -0,0 +1,138 @@
+import { useState, useEffect } from 'react';
+import Input from '../Input';
+import styles from './form.module.css';
+function Form() {
```
[the remark this answers is not in the document]
Comment (clementinob) [reply src/Components/Psychologists/Form/index.js:4]: Done 👍

File: src/Components/Psychologists/Form/index.js [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+  if (psychologistId) {
+    useEffect(() => {
+      fetch(`${process.env.REACT_APP_API}/psychologists/${psychologistId}`)
+        .then((response) => response.json())
+        .then((response) => {
+          setInputValues(response);
+        });
+    }, []);
+    options.method = 'PUT';
```
[the remark this answers is not in the document]
Comment (clementinob) [reply src/Components/Psychologists/Form/index.js:44]: You're right. Done! 😃

File: src/Components/Psychologists/DeleteBtn/index.js [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    })
+      .then((res) => {
+        if (res.status !== 200) {
+          return res.json().then((message) => {
+            throw new Error(message);
+          });
+        }
+      })
+      .catch((error) => error)
+      .finally(window.location.reload());
```
[the remark this answers is not in the document]
Comment (clementinob) [reply src/Components/Psychologists/DeleteBtn/index.js:26]: Yeap, my bad. Consider it done 👍

File: src/Components/Psychologists/Form/index.js [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+    })
+  };
+
+  if (psychologistId) {
+    useEffect(() => {
+      fetch(`${process.env.REACT_APP_API}/psychologists/${psychologistId}`)
+        .then((response) => response.json())
+        .then((response) => {
+          setInputValues(response);
+        });
```
[the remark this answers is not in the document]
Comment (clementinob) [reply src/Components/Psychologists/Form/index.js:42]: Implemented!

File: src/Components/Psychologists/Modal/modal.module.css [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+.modal {
+  width: 600px;
+  height: 300px;
+  background-color: white;
+  border-radius: 1px;
+}
+
+.modal button:disabled {
+	background: #555;
```
[the remark this answers is not in the document]
Comment (clementinob) [reply src/Components/Psychologists/Modal/modal.module.css:23]: ✔️

File: src/Components/Psychologists/Modal/modal.module.css [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+	cursor: not-allowed;
+}
+
+.modal button {
+  display:inline-block;
+  justify-content: center;
+  justify-content: space-around;
+  box-sizing: border-box;
+	width: 40%;
```
[the remark this answers is not in the document]
Comment (clementinob) [reply src/Components/Psychologists/Modal/modal.module.css:32]: ✔️

File: src/Components/Psychologists/Modal/modal.module.css (hunk 3 above)
Comment (clementinob) [reply src/Components/Psychologists/Modal/modal.module.css:36]: Done

File: src/Components/Psychologists/Modal/modal.module.css (hunk 4 above)
Comment (clementinob) [reply src/Components/Psychologists/Modal/modal.module.css:37]: ✔️

File: src/Components/Psychologists/Modal/modal.module.css (hunk 5 above)
Comment (clementinob) [reply src/Components/Psychologists/Modal/modal.module.css:49]: ✔️

File: src/Components/Psychologists/Modal/modal.module.css (hunk 6 above)
Comment (clementinob) [reply src/Components/Psychologists/Modal/modal.module.css:57]: ✔️

File: src/Components/Psychologists/psychologists.module.css (hunk 8 above)
Comment (clementinob) [reply src/Components/Psychologists/psychologists.module.css:24]: Done!

File: src/Components/Psychologists/Form/index.js [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+            throw new Error(ErrMessage);
+          });
+        }
+        return response.json();
+      })
+      .then(() => {
+        window.location.href = '/psychologists';
+      })
+      .catch((error) => {
+        console.log(error);
```
[the remark this answers is not in the document]
Comment (clementinob) [reply src/Components/Psychologists/Form/index.js:71]: Implemented!

File: src/Components/Psychologists/index.js (hunk 7 above)
Comment (clementinob) [reply src/Components/Psychologists/index.js:13]: Implemented!

**Review by nicolasviola — APPROVED** [review]

**Pull request merged**